### ISSUE TITLE
Thread deadlock on Bukkit

### ISSUE THREAD

**Issue by Packsolite** [LuckPerms/LuckPerms#3503]
### Description

Luckperms blocks the main thread causing the server to freeze and crash.

Issues seems to be related to https://github.com/LuckPerms/LuckPerms/issues/3483, although it is locking from a different call this time.

Attached a complete stacktrace dump to this issue.

### Reproduction Steps

Not quite sure as it is very rare, but does happen every now and then.

### Expected Behaviour

To not freeze the main thread.

### Server Details

Paperspigot-1.8.9-git-2894af0

### LuckPerms Version

5.4.49

### Logs and Configs

[stackdump.txt](https://github.com/LuckPerms/LuckPerms/files/9744335/stackdump.txt)

**Comment by LeotomasMC** [comment]
`[21:05:17] [Paper Watchdog Thread/ERROR]: 	eu.smashmc.smash.game.listener.PlayerToggleSneakListener.toggleSneak(PlayerToggleSneakListener.java:23)` whatever this plugin is, it appears to be checking permissions of lots of players whenever they sneak. report this to that plugin instead.

**Comment by Packsolite** [comment]
> `[21:05:17] [Paper Watchdog Thread/ERROR]: eu.smashmc.smash.game.listener.PlayerToggleSneakListener.toggleSneak(PlayerToggleSneakListener.java:23)` whatever this plugin is, it appears to be checking permissions of lots of players whenever they sneak. report this to that plugin instead.

This is our plugin, checking a permission once when a player happens sneaks under certain conditions.

**Comment by TechnicallyCoded** [comment]
Why are you not using `Player::hasPermission("perm")`? Luckperms already modifies the outcome of that.
If you are looking to avoid the bukkit api for whatever reason, why are you not using `User::getCachedData().getPermissionData().checkPermission("perm")`? There is no need to perform a lookup to a databse every time. Just use the already cached data.

**Comment by Packsolite** [comment]
> It looks like you are making queries to a database for this to happen? If so:
> 
> In general `Player::hasPermission("perm")` is safe? Luckperms already modifies the outcome of that. If you are looking to avoid the bukkit api for whatever reason, you can use `User::getCachedData().getPermissionData().checkPermission("perm")`? It shouldn't need to perform a DB query every time.
> 

I am using Player#hasPermission for this. The player is actually comming from a Bukkit.getOnlinePlayers() loop - so the player must be online and should already be cached by luckperms.

**Comment by TechnicallyCoded** [comment]
I think it would be useful if you would provide the entire log file following up to that crash, and you can choose to censor IPs if you wish to do so. I think also providing the 20-ish lines of code around your permissions check can help make sure there are no issues there.

**Comment by TechnicallyCoded** [comment]
Yeah I am not seeing anything obviously wrong here..
I was mostly interested in the logs though as that may have even small warnings at startup or after that would indicate issues with loading permissions for a player. If you are able to reproduce this behavior that would be great.

**Comment by LeotomasMC** [comment]
i highly suspect that "checking permsisions of many players when anyone crouches or uncrouches" has something to do with this problem. sure you have a cooldown, but you iterate over the blocks and check the permission on many players each iteration. there is no context here for how many iterations its doing on `blocks`. if you have a complex permissions setup with lots of parent groups, it *will* take longer to calculate permissions. being unable to see more code or details about how your permissions are setup, its hard to tell what the issue actually is.
i assume you are also setting and unsetting this `smash.advantage` permission in your plugin code. why not just store this advantage data yourself with a map or even on the player directly with bukkit's metadata?

**Comment by lucko** [comment]
LuckPerms doesn't actually appear the second time the server thread is dumped:

```
[21:05:17] [Paper Watchdog Thread/ERROR]: Current Thread: Server thread
[21:05:17] [Paper Watchdog Thread/ERROR]: 	PID: 39 | Suspended: false | Native: false | State: RUNNABLE
[21:05:17] [Paper Watchdog Thread/ERROR]: 	Thread is waiting on monitor(s):
[21:05:17] [Paper Watchdog Thread/ERROR]: 		Locked on:app//org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:475)
[21:05:17] [Paper Watchdog Thread/ERROR]: 	Stack:
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.util.HashMap.hash(HashMap.java:338)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.util.HashMap.getNode(HashMap.java:568)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.util.HashMap.getOrDefault(HashMap.java:1139)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.character.CharacterManager.getCharacter(CharacterManager.java:90)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.mechanic.VelocityHandler.getEffectiveVelocityMultiplier(VelocityHandler.java:121)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.mechanic.VelocityHandler.applyRawVelocity(VelocityHandler.java:99)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.character.characters.Toph.pushbackPlayers(Toph.java:163)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.character.characters.Toph.spawnWall(Toph.java:110)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.character.characters.Toph.onToggleSneak(Toph.java:49)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		eu.smashmc.smash.game.listener.PlayerToggleSneakListener.toggleSneak(PlayerToggleSneakListener.java:23)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		jdk.internal.reflect.GeneratedMethodAccessor77.invoke(Unknown Source)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.lang.reflect.Method.invoke(Method.java:568)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:300)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:74)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:62)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:490)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:475)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.PlayerConnection.a(PlayerConnection.java:1344)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.PacketPlayInEntityAction.a(SourceFile:43)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.PacketPlayInEntityAction.a(SourceFile:9)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:6)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.PlayerConnectionUtils$$Lambda$1485/0x00000008017c7208.run(Unknown Source)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.SystemUtils.a(SourceFile:44)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.MinecraftServer.B(MinecraftServer.java:781)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.DedicatedServer.B(DedicatedServer.java:376)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.MinecraftServer.A(MinecraftServer.java:720)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		app//net.minecraft.server.v1_8_R3.MinecraftServer.run(MinecraftServer.java:613)
[21:05:17] [Paper Watchdog Thread/ERROR]: 		java.base@17.0.3/java.lang.Thread.run(Thread.java:833)
```

Seems like a coincidence, but please feel free to open another issue if it occurs again.
